File: 3974-xor-after-range-multiplication-queries-i/xor-after-range-multiplication-queries-i.java
class Solution {
    public int xorAfterQueries(int[] nums, int[][] queries) {
        int MOD = 1_000_000_007;

        // Process each query
        for (int[] q : queries) {
            int l = q[0];
            int r = q[1];
            int k = q[2];
            int v = q[3];

            // Traverse indices: l, l+k, l+2k ...
            for (int idx = l; idx <= r; idx += k) {
                nums[idx] = (int)((1L * nums[idx] * v) % MOD);
            }
        }

        // Compute XOR of final array
        int xor = 0;
        for (int num : nums) {
            xor ^= num;
        }

        return xor;
    }
}

File: 3974-xor-after-range-multiplication-queries-i/xor-after-range-multiplication-queries-i.py
from typing import List

class Solution:
    def xorAfterQueries(self, nums: List[int], queries: List[List[int]]) -> int:
        MOD = 10**9 + 7
        
        # Process each query
        for l, r, k, v in queries:
            # range(start, stop, step) handles the ki jumps perfectly
            for i in range(l, r + 1, k):
                nums[i] = (nums[i] * v) % MOD
                
        # Calculate the final bitwise XOR
        ans = 0
        for num in nums:
            ans ^= num
            
        return ans